_posts/2022-06-21-qute-templates-from-db.adoc

```diff
@@ -0,0 +1,267 @@
+        }
+    }
+}
+----
+
+== Beware of Qute's internal cache
+
+Whenever Qute loads a template, it is stored into an internal `ConcurrentHashMap` and stays in memory forever, unless Qute is instructed otherwise.
+This means that you will need to remove a DB template from the Qute internal cache after it's been updated or deleted in the database.
+
+There are several ways of achieving that:
+
+[source,java]
+----
+package org.acme;
+
+import io.quarkus.qute.Engine;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+@ApplicationScoped
+public class DbEngineCacheManager {
+
+    @Inject
+    Engine engine;
+
+    public void removeTemplates(String name) {
+        engine.removeTemplates(templateName -> templateName.equals(name)); <1>
+    }
+
+    public void clearAll() {
+        engine.clearTemplates(); <2>
+    }
+}
+----
+<1> This removes the templates for which the mapping id matches the given predicate.
+<2> This removes all templates from the cache.
+
+Clearing that internal cache can become tricky if your app is running on a Kubernetes cluster with several replicas.
+You will indeed need a way to broadcast to all pods (possibly using a Kafka topic or a DB table) an instruction to remove from the cache the templates that have been updated or deleted.
+There is a cheaper (yet very imperfect) way of keeping all pods caches synced though, using a scheduled job:
+
+[source,java]
+----
+package org.acme;
+
+import io.quarkus.qute.Engine;
+import io.quarkus.scheduler.Scheduled;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+@ApplicationScoped
+public class DbEngineCacheScheduledCleaner {
+
+    @Inject
+    Engine engine;
+
+    @Scheduled(every = "5m", delayed = "5m") <1>
+    public void clearTemplates() {
+        engine.clearTemplates();
+    }
+}
+----
+<1> All templates will be cleared from the internal cache every 5 minutes.
+
+== Preventing the deletion of an included template
+
+A Qute template can be included into another template.
+If the inner template is deleted, then the outer template compilation will fail, which is obviously something that needs to be prevented while loading the templates from the DB.
+
+Here's a way to look for the inclusion of a template into another one before deleting it:
+
+[source,java]
+----
+package org.acme;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.transaction.Transactional;
+
+@ApplicationScoped
+public class TemplateRepository {
+
+    @Inject
+    EntityManager entityManager;
+
+    @Transactional
+    public void deleteTemplate(String name) {
+        long count = entityManager.createQuery("SELECT COUNT(*) FROM DbTemplate WHERE name != :name AND content LIKE :include", Long.class)
+                .setParameter("name", name)
+                .setParameter("include", "%{#include " + name + "%")
```

Comment:
A little late to the game, but isn't this line missing a `}`?

Reply:
Ah nevermind, it's a `LIKE` statement, I was looking at the parameter and forgot about the SQL above :)